File: application/backend-credit-card-enrollment/backend-c#/CreditCard/Enrollment/Query/GetUserEnrollmentsQueryHandler.cs
using CreditCardEnrollment.Common.Query;
using CreditCardEnrollment.Common.Projection;
using CreditCardEnrollment.Common.SessionAuth;
using CreditCardEnrollment.CreditCard.Enrollment.Projection.EnrollmentList;

namespace CreditCardEnrollment.CreditCard.Enrollment.Query;

public class GetUserEnrollmentsQueryHandler : QueryHandler {
    private readonly SessionService _sessionService;
    private readonly GetEnrollmentList _getEnrollmentList;

    public GetUserEnrollmentsQueryHandler(
        MongoTransactionalProjectionOperator mongoTransactionalProjectionOperator,
        SessionService sessionService,
        GetEnrollmentList getEnrollmentList) 
        : base(mongoTransactionalProjectionOperator) {
        _sessionService = sessionService;
        _getEnrollmentList = getEnrollmentList;
    }

    public override object HandleQuery(Common.Query.Query query) {
        if (query is GetUserEnrollmentsQuery enrollmentsQuery) {
            return HandleGetUserEnrollments(enrollmentsQuery);
        }
        
        throw new ArgumentException($"Unsupported query type: {query.GetType().Name}");
    }

    private IEnumerable<EnrollmentListItem> HandleGetUserEnrollments(GetUserEnrollmentsQuery query) {
        var userId = _sessionService.AuthenticatedUserIdFromSessionToken(query.SessionToken);
        return _getEnrollmentList.GetList(userId);
    }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/creditCard/enrollment/query/GetUserEnrollmentsQueryHandler.ts
import { QueryHandler } from '../../../common/query/QueryHandler';
import { GetUserEnrollmentsQuery } from './GetUserEnrollmentsQuery';
import {SessionService} from "../../../common/sessionAuth/SessionService";
import {EnrollmentListItem} from "../projection/enrollmentList/EnrollmentListItem";
import {GetEnrollmentList} from "../projection/enrollmentList/GetEnrollmentList";
import {MongoTransactionalProjectionOperator} from "../../../common/projection/MongoTransactionalProjectionOperator";
import {inject, injectable} from "tsyringe";

@injectable()
export class GetUserEnrollmentsQueryHandler extends QueryHandler {
    private readonly sessionService: SessionService;
    private readonly getEnrollmentList: GetEnrollmentList;

    constructor(
        @inject(MongoTransactionalProjectionOperator) mongoTransactionalProjectionOperator: MongoTransactionalProjectionOperator,
        @inject(SessionService) sessionService: SessionService,
        @inject(GetEnrollmentList) getEnrollmentList: GetEnrollmentList
    ) {
        super(mongoTransactionalProjectionOperator);
        this.sessionService = sessionService;
        this.getEnrollmentList = getEnrollmentList;
    }

    async handleQuery(query: GetUserEnrollmentsQuery): Promise<EnrollmentListItem[]> {
        const userId = await this.sessionService.authenticatedUserIdFromSessionToken(query.sessionToken);
        return this.getEnrollmentList.getList(userId);
    }
}
